sql/src/main/java/io/crate/operation/collect/collectors/NodeStatsIterator.java

```diff
@@ -95,53 +112,115 @@ private boolean isEmitDirectlyPossible(Set<ColumnIdent> toCollect) {
             transportStatTablesAction.execute(nodeId, request, new ActionListener<NodeStatsResponse>() {
                 @Override
                 public void onResponse(NodeStatsResponse response) {
                     rows.add(response.nodeStatsContext());
-                    if (remainingRequests.decrementAndGet() == 0) {
-                        emmitRows(rows);
+                    if (remainingNodesToCollect.decrementAndGet() == 0) {
+                        nodeStatsContextsFuture.complete(rows);
                     }
                 }
 
                 @Override
                 public void onFailure(Throwable t) {
                     if (t instanceof ReceiveTimeoutTransportException) {
                         rows.add(new NodeStatsContext(nodeId, node.name()));
-                        if (remainingRequests.decrementAndGet() == 0) {
-                            emmitRows(rows);
+                        if (remainingNodesToCollect.decrementAndGet() == 0) {
+                            nodeStatsContextsFuture.complete(rows);
                         }
                     } else {
-                        rowReceiver.fail(t);
+                        nodeStatsContextsFuture.completeExceptionally(t);
                     }
                 }
             }, TimeValue.timeValueMillis(3000L));
         }
+        return nodeStatsContextsFuture;
+    }
+
+    @Override
+    public Columns rowData() {
+        return rowData;
     }
 
-    private void emmitRows(List<NodeStatsContext> rows) {
-        new IterableRowEmitter(
-            rowReceiver,
-            RowsTransformer.toRowsIterable(inputFactory, RowContextReferenceResolver.INSTANCE, collectPhase, rows)
-        ).run();
+    @Override
+    public void moveToStart() {
+        raiseIfLoading();
+        it = rows.iterator();
+        currentRow = OFF_ROW;
+    }
+
+    @Override
+    public boolean moveNext() {
+        raiseIfLoading();
+        if (it.hasNext()) {
+            currentRow = it.next();
+            return true;
+        }
+        currentRow = OFF_ROW;
+        return false;
     }
 
     @Override
-    public void kill(@Nullable Throwable throwable) {
+    public void close() {
+    }
+
+    @Override
+    public CompletionStage<?> loadNextBatch() {
+        if (isLoading()) {
+            return CompletableFutures.failedFuture(new IllegalStateException("Iterator is already loading"));
+        }
+
+        if (allLoaded()) {
+            return CompletableFutures.failedFuture(new IllegalStateException("All batches already loaded"));
+        }
+
+        loading = new CompletableFuture<>();
+
+        CompletableFuture<List<NodeStatsContext>> nodeStatsContexts = getNodeStatsContexts();
+        nodeStatsContexts.whenComplete((List<NodeStatsContext> result, Throwable t) -> {
+            if (t == null) {
+                rows = RowsTransformer.toRowsIterable(inputFactory, RowContextReferenceResolver.INSTANCE, collectPhase,
```

Comment:
it would be nice to get rid of this intermediate row transformation, however we can to this in a sepereate step, since there are already other implementations in the target branch using this still too